### PR TITLE
chapter 3 implement

### PR DESCRIPTION
**Pull request by Dondo722**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/service/DefaultCartService.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class DefaultCartService implements CartService{
+    private static final String CART_SESSION_ATTRIBUTE = DefaultCartService.class.getName() + ".cart";
+    private static final Object instanceLock = new Object();
+    private static volatile DefaultCartService instance;
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
```
Comment (Dondo722) [inline src/main/java/com/es/phoneshop/service/DefaultCartService.java:19]: i don't like lockers here, the lecturer suggested to use  session lock here, but i found only this 
https://stackoverflow.com/questions/9802165/is-synchronization-within-an-httpsession-feasible
but what stopped me, is calculating session lock. Is it good solution or not?

File: src/main/webapp/WEB-INF/pages/product.jsp [hunk 2]
```diff
@@ -5,27 +5,76 @@
...
+                ${error}
+              </div>
+            </c:if>
+          </td>
+        </tr>
       </tr>
-  </table>
+    </table>
+    <button>Add to card</button>
+    <c:if test="${not empty viewed.history}">
```
Comment (Dondo722) [inline src/main/webapp/WEB-INF/pages/product.jsp:57]: it is code duplication i tried make tag out of this block<c:if> </c:if>, with viewed as attribute, but it didn't work. 
Should I keep trying this way?

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 3]
```diff
@@ -1,33 +1,85 @@
...
-        } catch (NumberFormatException e) {
+        } catch (ParseProductFromRequestException e) {
+            response.sendError(500);
+        }
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        try {
+            String quantityString = request.getParameter("quantity");
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:54]: I was able to use negative values here to add an item to cart like: -1.

File: src/main/java/com/es/phoneshop/service/DefaultViewedService.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            Product cachedProduct = viewedHistory.getCache();
+            if (cachedProduct != product) {
+                viewedHistory.setCache(product);
+                if (cachedProduct == null) {
+                    return;
+                }
+                if (!viewed.contains(cachedProduct)) {
+                    viewed.add(0, cachedProduct);
+                }
+                if (viewed.size() > 3) {
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/service/DefaultViewedService.java:67]: I think 3 should be moved to constant for better visibility.

File: src/main/java/com/es/phoneshop/service/DefaultCartService.java (hunk 1 above)
Comment (SergeiStarikov) [reply src/main/java/com/es/phoneshop/service/DefaultCartService.java:19]: It's really enough to use synchronized method/block here.
Locks in this lab are mostly to get acquainted with various synchronization techniques.

**Review by SergeiStarikov — APPROVED** [review]

File: src/main/webapp/WEB-INF/pages/product.jsp (hunk 2 above)
Comment (SergeiStarikov) [reply src/main/webapp/WEB-INF/pages/product.jsp:57]: It you have time, you can try more to extract the code into a tag. If not, it's fine to have some code duplication here.

**Pull request merged**